File: 476.number-complement.js
/*
 * @lc app=leetcode id=476 lang=javascript
 *
 * [476] Number Complement
 */

// @lc code=start
/**
 * @param {number} num
 * @return {number}
 */
const findComplement = function (num)
{
    let mask = 1;
    for (let i = 0; i < 32; i++)
    {
        if (num >= mask)
        {
            mask *= 2;
        }
        else
        {
            mask--;
            break;
        }
    }
    return (~num) & mask;
};
// @lc code=end

File: Daily/476.number-complement.cpp
/*
 * @lc app=leetcode id=476 lang=cpp
 *
 * [476] Number Complement
 */

// @lc code=start
class Solution {
 public:
  int findComplement(int num) {
    int result = 0;
    for (int i = 0; num > 0; i++, num >>= 1) {
      const int bit = num & 0b1;
      result += (0b1 - bit) << i;
    }
    return result;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.findComplement(5);
}
